### ISSUE TITLE
acme: self-check http requests can't handle https redirects

### ISSUE THREAD

**Issue by tinnus-napbus** [urbit/urbit#6234]
**Describe the bug**
initial cert configuration seemingly works, but 3 months later the self-checks fail during renewal because acme can't handle the now-enabled http->https redirects in eyre

```
acme: unable to retrieve self-hosted domain validation token via xxxxxxxxxxxxx.arvo.network. please confirm your urbit has network connectivity.
[ p=301
    q
  ~[
    [p='date' q='Sun, 22 Jan 2023 03:46:31 GMT']
    [p='connection' q='close']
    [p='server' q='urbit/vere-1.15']
    [ p='location'
        q
      'https://xxxxxxxxxxxxxxxxx.arvo.network/.well-known/acme-challenge/xxxxxxxxxxxxxxxxxxxxxxxxx'
    ]
  ]
  r=~
]
```

**To Reproduce**
Steps to reproduce the behaviour:
1. run `-dns-address` thread & configure domain & ssl certs
2. wait until renewal
3. errors during self-check because of 301 redirects

**Expected behaviour**
should handle 301 redirects

**System (please supply the following information, if relevant):**
 - OS: aarch64 linux
 - Vere: 1.15 & 2ucn6

**Comment by kingdonb** [comment]
I've been experiencing this issue as well, I've been trying to understand the shape of it

As far as I can tell it's now Urbit itself that serves up the 301 redirect, and the HTTP-01 response is only served up via TLS, which is not conventional unless you're doing an ALPN challenge as far as I know. It should be served on the http port.

I too have routers in front of my Urbit ship, it doesn't answer directly, but each port goes directly to Urbit:
https://github.com/kingdonb/bootstrap-repo/blob/staging/apps/routers/urbit-ships/socryx-topled-ingressroute.yaml

https://github.com/kingdonb/bootstrap-repo/blob/staging/apps/routers/urbit-ships/socryx-topled-service-endpoints.yaml#L26

The backend is running urbit on this IP, port 80 and 443 – for now each ship in my "backend" will get its own physical node to accommodate the structure, but the point is, I'm almost certain this is a regression in Urbit's `%acme` because it has always worked before. (But I'm sure you knew that, after re-reading your report you're more in the loop than I am...)

Funnier thing is, my TLS certificate doesn't even lose the solid lock icon – but if you inspect it, the cert has clearly expired some time ago. I'm not sure why this cert is apparently considered as valid even after it has expired.

**Comment by tacryt-socryp** [comment]
I tried to reset my `acme` and wasn't able to, even when I nuked it.

**Comment by kingdonb** [comment]
as far as I know, HTTP-01 challenges must be advertised on the HTTP port. They will not work on port 443.

So if eyre is redirecting everything from port 80 to 443, it's going to need to make an exception for the HTTP-01 LE challenge.

**Comment by tacryt-socryp** [comment]
How can I just turn off Eyre HTTPS so I can try to retry Acme?

**Comment by tinnus-napbus** [comment]
> as far as I know, HTTP-01 challenges must be advertised on the HTTP port. They will not work on port 443.

they must be initially available on port 80 but the letsencrypt bot will follow redirects.

> How can I just turn off Eyre HTTPS so I can try to retry Acme?

`|pass [%e %rule %cert ~]` will disable https.

**Comment by tacryt-socryp** [comment]
Thanks, was able to fix my issue by using that command.

**Comment by galenwp** [comment]
That also got me into a state that works—but I do think `acme` is broken for the time being. 

(And the funny thing is I'm getting even more confusing errors from Let'sEncrypt's certbot on the box I'm running on.)

**Comment by kingdonb** [comment]
I've kept my pier in a broken state in case there's help needed to test a fix

I'm not sure if this affects everyone, but it seems like it might affect everyone who has configured acme, (and that might be a short list.)

There is a redirect on the challenge URL, it seems to be follow-able, but the challenge is not accepted or acme is not seeing that it gets accepted. I'm not sure where it gets hung up. Here is the verbose output from curl with redirect follow enabled.

```
$ curl -vL http://socryx-topled.arvo.network/.well-known/acme-challenge/4JFwP4ltavKJbzYuuK5V9zNJKfFz1s_qLLKA4b4J_40
*   Trying 98.227.195.144:80...
* Connected to socryx-topled.arvo.network (98.227.195.144) port 80 (#0)
> GET /.well-known/acme-challenge/4JFwP4ltavKJbzYuuK5V9zNJKfFz1s_qLLKA4b4J_40 HTTP/1.1
> Host: socryx-topled.arvo.network
> User-Agent: curl/7.85.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 301 Moved Permanently
< Date: Tue, 24 Jan 2023 16:49:19 GMT
< Location: https://socryx-topled.arvo.network/.well-known/acme-challenge/4JFwP4ltavKJbzYuuK5V9zNJKfFz1s_qLLKA4b4J_40
< Server: urbit/vere-1.17
< Transfer-Encoding: chunked
<
* Ignoring the response-body
* Connection #0 to host socryx-topled.arvo.network left intact
* Clear auth, redirects to port from 80 to 443
* Issue another request to this URL: 'https://socryx-topled.arvo.network/.well-known/acme-challenge/4JFwP4ltavKJbzYuuK5V9zNJKfFz1s_qLLKA4b4J_40'
*   Trying 98.227.195.144:443...
* Connected to socryx-topled.arvo.network (98.227.195.144) port 443 (#1)
* ALPN: offers h2
* ALPN: offers http/1.1
*  CAfile: /etc/ssl/cert.pem
*  CApath: none
* (304) (OUT), TLS handshake, Client hello (1):
* (304) (IN), TLS handshake, Server hello (2):
* (304) (IN), TLS handshake, Unknown (8):
* (304) (IN), TLS handshake, Certificate (11):
* (304) (IN), TLS handshake, CERT verify (15):
* (304) (IN), TLS handshake, Finished (20):
* (304) (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / AEAD-CHACHA20-POLY1305-SHA256
* ALPN: server accepted h2
* Server certificate:
*  subject: CN=socryx-topled.arvo.network
*  start date: Oct 31 12:01:50 2022 GMT
*  expire date: Jan 29 12:01:49 2023 GMT
*  subjectAltName: host "socryx-topled.arvo.network" matched cert's "socryx-topled.arvo.network"
*  issuer: C=US; O=Let's Encrypt; CN=R3
*  SSL certificate verify ok.
* Using HTTP2, server supports multiplexing
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* h2h3 [:method: GET]
* h2h3 [:path: /.well-known/acme-challenge/4JFwP4ltavKJbzYuuK5V9zNJKfFz1s_qLLKA4b4J_40]
* h2h3 [:scheme: https]
* h2h3 [:authority: socryx-topled.arvo.network]
* h2h3 [user-agent: curl/7.85.0]
* h2h3 [accept: */*]
* Using Stream ID: 1 (easy handle 0x13a013600)
> GET /.well-known/acme-challenge/4JFwP4ltavKJbzYuuK5V9zNJKfFz1s_qLLKA4b4J_40 HTTP/2
> Host: socryx-topled.arvo.network
> user-agent: curl/7.85.0
> accept: */*
>
< HTTP/2 200
< server: urbit/vere-1.17
< date: Tue, 24 Jan 2023 16:49:20 GMT
< content-type: text/html
< content-length: 87
<
* Connection #1 to host socryx-topled.arvo.network left intact
4JFwP4ltavKJbzYuuK5V9zNJKfFz1s_qLLKA4b4J_40.NRP8wTzbQQLgY1KjnSAxKshxd8oef23nk6fiPgHga28
```

**Comment by rabsef-bicrym** [comment]
@tinnus-napbus FYI I had to perform the `eyre` reset today to get my cert re-authenticated (`|pass [%e %rule %cert ~]`).

```
> +vat %base
%base
  /sys/kelvin:      [%zuse 415]
  base hash:        0v3.sgh93.3pomr.cup26.rivn8.0g8nj.fs6d4.n48ji.5q8g5.r1niq.h0at8
  %cz hash:         0v18.sqdbi.r6vna.quaq0.drt53.pbhhg.jcfn9.svdnj.shn20.vb414.kjsgt
  app status:       running
  force on:         ~
  force off:        ~
  publishing ship:  ~
  updates:          remote
  source ship:      ~litzod
  source desk:      %kids
  source aeon:      131
  kids desk:        %kids
  pending updates:  ~
  ```
  
  I had modified acme in my base desk to follow what was going on (sigpaming out some things) and saw that I was in the validate-domain loop. AS a last resort, performing that reset fixed things, still.

**Comment by matildepark** [comment]
♫ thank you rabsef

I had the same issue today.

**Comment by rabsef-bicrym** [comment]
Incidentally, this breaks access to Scene.

**Comment by matildepark** [comment]
This would break access to anything. The SSL cert is invalid.

**Comment by rabsef-bicrym** [comment]
sure, but in browser I can ignore that but not in my preferred mode.

**Comment by chc4** [comment]
This issue should not be closed, because the issue isn't actually fixed. LetsEncrypt self-checks are still broken, and so certificates aren't renewing without a manual `|pass [%e %rule %cert ~]` to disable SSL every 3 months, which sucks. `:acme` should auto-renew the certificate, and it isn't because Vere doesn't serve HTTP correctly.

**Comment by rabsef-bicrym** [comment]
>  Vere doesn't serve HTTP correctly.

Could you elaborate for me? This interests me but I wasn't aware there was a problem.

**Comment by vvisigoth** [comment]
@chc4 @joemfb I'd like to try to get a bounty out to fix some of these underlying HTTP problems. Can either of you summarize what the root problem is here and what can/should be done to ameliorate?

**Comment by pkova** [comment]
My understanding of the lifecycle of this bug is the following:

In the beginning before we had http->https redirects everything renewals were working fine. When the redirects were added mysterious errors referencing HTTP 301 started appearing, even though letsencrypt follows redirects completely fine. Iris, however, does not! Acme does a self check which was now broken, but @tinnus-napbus fixed this in #6238. 

However, this line https://github.com/urbit/urbit/blob/b271d5e3c33e19271e4b31e96bda8c7765e22606/pkg/arvo/app/acme.hoon#L589
where we scry eyre to check whether we're secure or not was just returning the wrong result. This was fixed again by @tinnus-napbus in #6116, which was released in [`urbit-os-v2.133`](https://github.com/urbit/urbit/releases/tag/urbit-os-v2.133) to the network today.

TLDR I'm pretty sure this issue is now fixed. I don't have any affected ships to verify, though.

**Comment by vvisigoth** [comment]
@pkova Me and a few people in the support emails still have unrenewed certs. How can we check if this fix worked?

**Comment by pkova** [comment]
My suspicion is that if your certificate expired while this bug was out and about `%acme` will have ended up in a bad state. The way to instantly fix it is to reset by running `|pass [%e %rule %cert ~]` followed by `:acme &path /network/arvo/sampel-palnet`. Note that the path in `:acme` is reversed from the actual domain, for example my url `dinleb-rambep.arvo.network` becomes `/network/arvo/dinleb-rambep`.

I'm pretty sure that people who had valid certs when the fix went out yesterday will be fine.

**Comment by kingdonb** [comment]
```
http: fail (51, 504): certificate has expired
> |pass [%e %rule %cert ~]
>=
http: restarting servers to apply configuration
http: web interface live on http://localhost:80
http: loopback live on http://localhost:12321
> :acme /network/arvo/socryx-topled
>=
kiln: downloading update for %sentinel from ~dister-dozzod-lapdeg/%sentinel
kiln: finished downloading update for %sentinel from ~dister-dozzod-lapdeg/%sentinel
kiln: remote is identical to %sentinel from ~dister-dozzod-lapdeg/%sentinel, skipping
> :acme &path /network/arvo/socryx-topled
>=
acme: requesting an https certificate for socryx-topled.arvo.network
acme: received https certificate for socryx-topled.arvo.network
http: restarting servers to apply configuration
http: web interface live on https://localhost:443
http: web interface live on http://localhost:80
http: loopback live on http://localhost:12321
```

Looks like it worked. Thank you!

**Comment by lukestiles** [comment]
This bug may have returned. I'm trying to update ~roslet-tanner and getting an error 

> |pass [%e %rule %cert ~]
>=
http: restarting servers to apply configuration
http: web interface live on http://localhost:8080
http: loopback live on http://localhost:12321
> |pass [%e %rule %cert ~]
>=
> :acme &path /network/arvo/roslet-tanner
>=
acme: requesting an https certificate for roslet-tanner.arvo.network
http: fail (6, 504): connection failure
http: fail (7, 504): connection failure
http: fail (8, 504): connection failure
%group-store: on-poke with mark %group-update-0
/groups
http: fail (9, 504): connection failure
; ~zod not responding still trying
; ~zod is ok
ames: czar bus.urbit.org: ip .35.247.126.229
http: fail (10, 504): connection failure

**Comment by tinnus-napbus** [comment]
@lukestiles your ship is running on port 8080, it needs to be running on port 80. If you're on linux you probably need to do `sudo setcap 'cap_net_bind_service=+ep' <pier>/.run` to allow it to bin port 80/443.

Once that's done & you restart your ship and it says `http: web interface live on http://localhost` or localhost:80 or w/e, try again and hopefully it'll work

**Comment by lukestiles** [comment]
> @lukestiles your ship is running on port 8080, it needs to be running on port 80. If you're on linux you probably need to do `sudo setcap 'cap_net_bind_service=+ep' <pier>/.run` to allow it to bin port 80/443.
> 
> Once that's done & you restart your ship and it says `http: web interface live on http://localhost` or localhost:80 or w/e, try again and hopefully it'll work

That did it. Thank you.

**Comment by ngzax** [comment]
I can't connect to ~winter-paches at all. Tried the above but still no connecto:

```
> |pass [%e %rule %cert ~]
>=
http: restarting servers to apply configuration
http: web interface live on http://localhost:80
http: loopback live on http://localhost:12321
%posh-fail
> :acme /network/arvo/winter-paches
>=
> :acme &path /network/arvo/winter-paches
>=
acme: requesting an https certificate for winter-paches.arvo.network
acme: received https certificate for winter-paches.arvo.network
http: restarting servers to apply configuration
http: web interface live on https://localhost:443
http: web interface live on http://localhost:80
http: loopback live on http://localhost:12321
```

All my other ships are fine. :(

**Comment by tinnus-napbus** [comment]
@ngzax I get to its login screen at winter-paches.arvo.network so it seems to be working now

**Comment by ngzax** [comment]
Well that's kind of strange but it's working for me now as well. ¯\_(ツ)_/¯

On Apr 3, 2023, at 6:35 PM, tinnus-napbus ***@***.***> wrote:


@ngzax I get to its login screen at winter-paches.arvo.network so it seems to be working now

**Comment by ngzax** [comment]
aaaand it's failing again.

<img width="1101" alt="image" src="https://user-images.githubusercontent.com/87197/230106601-1845d351-826c-4df6-8a5d-ebaea1dddcd3.png">

**Comment by ngzax** [comment]
scrolling up on the console i do see this which looks worrisome. note that i had done nothing in the browser since it was last working. just went and refreshed and boom, not working.

```
crud: %ames-wake event failed
[%poke %wake]
wire=/pump/~pathus-hiddyn/759
bar-stack=[i=[i=/ames t=~] t=[i=[i=//behn t=~] t=~]]
take: failed
/sys/vane/ames/hoon:<[1.232 3].[1.252 20]>
/sys/vane/ames/hoon:<[1.233 3].[1.252 20]>
/sys/vane/ames/hoon:<[1.237 3].[1.252 20]>
/sys/vane/ames/hoon:<[1.239 3].[1.252 20]>
/sys/vane/ames/hoon:<[1.240 5].[1.250 7]>
/sys/vane/ames/hoon:<[1.241 5].[1.250 7]>
/sys/vane/ames/hoon:<[1.245 24].[1.245 65]>
/sys/vane/ames/hoon:<[2.040 5].[2.088 55]>
/sys/vane/ames/hoon:<[2.042 5].[2.088 55]>
/sys/vane/ames/hoon:<[2.056 5].[2.088 55]>
/sys/vane/ames/hoon:<[2.057 7].[2.069 71]>
/sys/vane/ames/hoon:<[2.058 7].[2.069 71]>
/sys/vane/ames/hoon:<[2.062 7].[2.069 71]>
/sys/vane/ames/hoon:<[2.063 7].[2.069 71]>
/sys/vane/ames/hoon:<[2.068 7].[2.069 71]>
/sys/vane/ames/hoon:<[2.069 7].[2.069 71]>
/sys/vane/ames/hoon:<[2.069 12].[2.069 71]>
/sys/vane/ames/hoon:<[2.759 7].[2.828 38]>
/sys/vane/ames/hoon:<[2.762 7].[2.828 38]>
/sys/vane/ames/hoon:<[2.783 7].[2.828 38]>
/sys/vane/ames/hoon:<[2.791 7].[2.828 38]>
/sys/vane/ames/hoon:<[2.808 7].[2.828 38]>
/sys/vane/ames/hoon:<[2.820 7].[2.828 38]>
/sys/vane/ames/hoon:<[2.828 7].[2.828 38]>
/sys/vane/ames/hoon:<[2.885 7].[3.003 9]>
/sys/vane/ames/hoon:<[2.888 7].[3.003 9]>
/sys/vane/ames/hoon:<[2.891 7].[3.003 9]>
/sys/vane/ames/hoon:<[2.892 7].[3.003 9]>
/sys/vane/ames/hoon:<[2.893 7].[3.003 9]>
/sys/vane/ames/hoon:<[2.893 47].[2.893 71]>
/sys/vane/ames/hoon:<[3.225 5].[3.232 7]>
/sys/vane/ames/hoon:<[3.227 5].[3.232 7]>
/sys/vane/ames/hoon:<[3.227 9].[3.227 29]>
/sys/vane/ames/hoon:<[3.237 5].[3.273 7]>
/sys/vane/ames/hoon:<[3.239 5].[3.273 7]>
/sys/vane/ames/hoon:<[3.242 14].[3.242 51]>
/sys/vane/ames/hoon:<[3.433 5].[3.438 50]>
/sys/vane/ames/hoon:<[3.435 5].[3.438 50]>
/sys/vane/ames/hoon:<[3.436 7].[3.436 42]>
/sys/vane/ames/hoon:<[3.491 5].[3.501 7]>
/sys/vane/ames/hoon:<[3.493 5].[3.501 7]>
/sys/vane/ames/hoon:<[3.495 5].[3.501 7]>
/sys/vane/ames/hoon:<[3.498 14].[3.498 36]>
/sys/vane/ames/hoon:<[3.526 5].[3.565 50]>
/sys/vane/ames/hoon:<[3.529 5].[3.565 50]>
eyre: canceling [i=//http-server/0v3.60old/418/1 t=~]
ames: czar at lur.urbit.org: not found (b)
ames: czar at bus.urbit.org: not found (b)
error in %chat
%poke-ack
! take %poke-ack failed
```

**Comment by tinnus-napbus** [comment]
@ngzax the ames error is unrelated. I can't tell whether it's a DNS resolution error or something else from the screenshot you posted. You could possibly try deleting the certificate again and re-running the `-dns-address` thread with your ip address in case the dns record has gotten weird somehow. I can still access your ship's login screen though so idk, it seems like the record is correct. Maybe you should try clearing browser caches etc in case that's in a weird state?

**Comment by a-moreira** [comment]
I have just encountered the certificate error while re-starting a planet that was offline for a couple of months. I can confirm that
```
|pass [%e %rule %cert ~]
```
followed by
```
:acme &path /network/arvo/<planet>
```
gets it running with TLS again.

**Comment by kingdonb** [comment]
I'm currently unable to renew my cert. I dumped it with `|pass [%e %rule %cert ~]` and restarted the pier as root (macos) so it only listens on port 80 now. This didn't help.

I also had some janky forwarding maze with Traefik in front of the planet that I thought might have been a contributing factor. I changed it to the simplest possible NAT forwarding rule for port 80 directly to the planet host, and that did not help.

The error:

```
> :acme %retry
>=
acme: requesting an https certificate for socryx-topled.arvo.network
acme: unable to reach ~socryx-topled via http at socryx-topled.arvo.network:80
[ p=500
    q
  ~[
    [p='date' q='Fri, 28 Jun 2024 15:32:55 GMT']
    [p='connection' q='keep-alive']
    [p='content-length' q='174']
    [p='server' q='urbit/vere-3.0']
    [ p='set-cookie'
        q
      'urbauth-~socryx-topled=0vq2fjs.4lefq.66606.lbj2v.outil; Path=/; Max-Age=604800'
    ]
    [p='content-type' q='text/html']
  ]
    r
  [ ~
    [ p=174
        q
      26.354.964.062.611.594.843.583.896.911.973.069.817.834.363.092.271.408.532.914.035.607.136.008.912.921.123.832.810.867.626.287.095.865.698.641.497.932.215.489.078.432.061.211.289.681.856.143.573.665.980.773.702.224.885.720.521.849.860.754.995.394.108.101.706.748.803.508.702.719.073.188.649.067.554.840.287.180.655.324.428.230.453.701.446.628.507.900.631.428.922.407.129.476.044.828.216.141.103.989.275.953.101.257.669.176.822.598.009.067.585.128.940.623.302.565.892.229.710.817.071.220.207.285.541.757.545.121.982.285.648.850.793.760.949.635.839.727.790.140
    ]
  ]
]
error in %acme
%arvo-response
! /app/acme/hoon:<[579 45].[579 47]>
! %no-next-order
! /app/acme/hoon:<[579 27].[579 48]>
! /app/acme/hoon:<[579 5].[596 40]>
! %validate-domain-effect-fail
...
```

This error is strange because I can see no reason why acme would be unable to reach itself at that address. The pier is running on my laptop, but I can get through to that address with a browser from inside or outside of the local network, no problemo. If I hit it on port 80 with curl, I can see I am getting redirected to login, and the redirect is successful if followed.

I am not sure how to inspect the challenge/order.

```
acme: unable to reach ~socryx-topled via http at socryx-topled.arvo.network:80
```

**Comment by kingdonb** [comment]
Absolutely no idea what changed but now I have a valid cert.

```
acme: received https certificate for socryx-topled.arvo.network
http: restarting servers to apply configuration
http: web interface live on https://localhost:443
http: web interface live on http://localhost:80
http: loopback live on http://localhost:12321
```

**Comment by kingdonb** [comment]
In case it's illustrative about how this error can still be raised, here are the Kubernetes service and endpoints config for my Urbit planet, and the front-end Traefik IngressRoute/IngressRouteTCP that enables me to do this. These routes were in the path when the cert was provisioned successfully. The only difference was that `targetPort` settings were changed from 8080/8443 to the port 80/443 to match the Urbit running as root.

I guess this isn't an issue on Linux where you can `setcap +ep`.

**Comment by kingdonb** [comment]
I'm renewing my cert today, and this doc is coming in handy... maybe I better blog about it.

**Comment by tinnus-napbus** [comment]
If you are having issues with cert renewal, note there is a bug as described here affecting it https://github.com/urbit/urbit/pull/7059 

You may need to wait until that fix is released or else apply the patch yourself manually if you are able

**Comment by kingdonb** [comment]
Thanks for the information. I am a planet (not comet) so I did not have that issue.

```
acme: received https certificate for socryx-topled.arvo.network
http: restarting servers to apply configuration
http: web interface live on https://localhost:443
http: web interface live on http://localhost:80
```

The planet has to run as root to receive an SSL certificate, last time I tried it anyway it did, because the redirect is using the planet's own ports. So a planet running on port 8080 / 8443 will attempt to redirect the port 8080 request to port 8443.

That is, even though I am using the router config described in:
* https://github.com/urbit/urbit/issues/6234#issuecomment-2197640698

which serves up the HTTP/HTTPS ports on the front-end on a load balancer / host-based SNI router using ports 80 and 443, the redirect from 80 goes to port 8443, which isn't listening, and causes the whole thing to fail.

This is worked around by running my planet as `sudo ./socryx-topled/.run -p 60007 -b 10.17.12.102` just once, while the router config has been updated like this:

```
    ports:
    - name: urbit-https
      protocol: TCP
      port: 443
      targetPort: 443
      nodePort: 0
    - name: urbit-http
      protocol: TCP
      port: 80
      targetPort: 80
      nodePort: 0
```

It can be reverted after the certificate is received. It might make sense for the front-end router to use port 80/443 AND port 8080/8443 so this dance isn't required in the future. I will definitely bookmark this issue, already have it bookmarked, and I will try that next time, perhaps update again in ~60 days or when there is time to boot a secondary planet, (not sure how the VM could ever reasonably be expected to understand that its own port 8080 is really port 80, and 8443 is really port 443.)

Then I think the planet I'm hosting on Kubernetes wouldn't ever need to be given root or cap_net_bind_service, or any of the privileged bits.

**Comment by tinnus-napbus** [comment]
It doesn't matter what rank your ship actually is. Unauthenticated HTTP requests come in as a random comet and the self-check doesn't bother with authentication. Also, I think you might have issues with the self-check due to your reverse-proxy but I'm not 100% sure. I am looking at ripping out the self-check stuff or making it optional because it can cause issues in certain configurations

**Comment by kingdonb** [comment]
It must have worked because I already have a valid certificate then, I'm just renewing it.

**Comment by kingdonb** [comment]
I'll wait for a release containing #7059 before I try a new planet then, to avoid running up against that issue.

(is that an arvo release I'd be waiting on, where should I look for that? I am on urbit-dev and I receive the `ANNOUNCE` mails, so I'd guess to watch for it there)

**Comment by tinnus-napbus** [comment]
When an SSL cert is configured, I'd have thought the urbit binary would bind 8080 and 8443 separately. It definitely does when it's been privileged with setcap or w/e to bind 80 and 443. I'm surprised insecure traffics works through the https port but actually I haven't tried it so maybe it does.

NAT hairpinning is when you make a request to your public IP address from within your LAN, and your router routes it back to the LAN again based on its address translation table. Not all consumer routers do it, though I'm not clear if you're running this at home or on a VPS (VPSes will basically always do it).

Note that rather than running your ship with sudo to allow it to bind 80/443, you can also do `sudo setcap 'cap_net_bind_service=+ep' <pier>/.run` to give it that privilege. If you actually need the reverse-proxy though you probably wouldn't want to as they'd conflict

**Comment by kingdonb** [comment]
No, insecure traffic does not work on the https port. I'm running in kubernetes where I really prefer not to add privileges or rely on setcap because it complicates things significantly, even compared to all this.

The urbit binary binds port 8080 and port 8443, the kubernetes service forwards from 8080 to 8080 and 8443 to 8443, then in front of that, a traefik router creates entrypoints "web" and "webnonpriv" which are listening on 8000/8080, and ports "websecure" and "websecnonpriv", on 8443 and 8445 (because again, the deployment shouldn't have privileges)

Then finally a frontend LoadBalancer on port 80, 8080, 443, and 8443, forwards to those entrypoints, which get through on the backend to 8080 and 8443 just fine. The only weirdness is that redirects always go to 8443. Port 443 doesn't "302 redirect" but it does translate to the backend service on port 8443, opaquely to the user, which is perfect.

I've been iterating this config for literally years. I think now it's basically perfect.

Here is the ingressroute / service / endpoints config: (the ship itself does not run in kubernetes, so endpoints are written by hand instead of with service / pod selectors)

* https://github.com/kingdonb/bootstrap-repo/tree/main/apps/routers/urbit-ships

Here is the traefik config for kubernetes:

* https://github.com/kingdonb/bootstrap-repo/tree/main/apps/staging/traefik
* (which derives from: https://github.com/kingdonb/bootstrap-repo/tree/main/apps/base/traefik)

The loadbalancer defines all ports, 80/8080/443/8443:

* https://github.com/kingdonb/bootstrap-repo/blob/e71519a6e59f9aa2ef16ce96d86ed5f9241b333b/apps/base/traefik/svc.yaml#L17-L32

The service and deployment backends only define port 8080 and 8443:

* https://github.com/kingdonb/bootstrap-repo/blob/e71519a6e59f9aa2ef16ce96d86ed5f9241b333b/apps/routers/urbit-ships/socryx-topled-ingressroute.yaml
* https://github.com/kingdonb/bootstrap-repo/blob/e71519a6e59f9aa2ef16ce96d86ed5f9241b333b/apps/routers/urbit-ships/socryx-topled-service-endpoints.yaml#L38-L41

(the current service config defines also 80/443 ports but I think they aren't used, they will be cleaned up later)

Before I try to publish this anywhere I will isolate it from all my other examples and put it in a solo-repo for just this example.

**Comment by kingdonb** [comment]
I talked the cozystack maintainers into doing a similar thing with their ingress machinery, the idea is generally to be able to support deployments of backends that are responsible for their own TLS termination, their own letsencrypt ACME HTTP-01 challenge orders, they should be able to respond to port 80 and port 443 independently, but Kubernetes really does not want you to build this config with the standard `Ingress` which is why I resorted to using Traefik, when I normally use ingress-nginx - the Traefik IngressRoute and IngressRouteTCP CRDs support these things very flexibly, (where K8s normal Ingress can only have one single backend port, and the front-end router does its own redirect precluding any of this from working, it's meant to terminate TLS for you... No good!)

**Comment by tinnus-napbus** [comment]
ok yeah I get it now. I wrote the redirect logic to append the port number that urbit has configured for SSL (in your case 8443) to the rewritten URL if it's not the standard 443 port. Urbit can't know about the stuff in front of it and if it didn't do this it wouldn't work if you were actually raw-dogging it on 8443 or w/e. Idk how to fix this except perhaps to add a task to Eyre to disable the redirect behaviour. Currently it's mandatory when SSL is enabled because auto-redirect is the default & preferred behaviour on 99% of the web, and without it noobs would unknowingly access their web interface insecurely. I will look into how practical changing this would be as I'm working on some related stuff anyway

**Comment by tinnus-napbus** [comment]
Most people either raw-dog urbit or else have it behind a reverse-proxy that handles letsencrypt for them, so I've not really dealt with this case of having urbit behind stuff while still handling its own SSL certs before

**Comment by kingdonb** [comment]
Thanks, I'm glad that makes sense. I don't think there's any way to fix it as you say Urbit cannot know what's in front. This doesn't really feel like a footgun, or nonsense, it's just a bit complex but it really works at this point, doesn't feel like anything incorrect, so I don't know if we need to fix it.

But for anyone who wants to build a hosting platform it will be helpful to have a clear documentation of all this, maybe that's something I can work on outside of the Urbit proper.

But perhaps a special config to expect a loadbalancer would make sense. So users could opt-in to have the redirect to port 443 instead of the 8443 that it's actually listening on.

**Comment by kingdonb** [comment]
So, I've got my networking configured in such a way that port 80 and 443 are used in the load balancer, but port 8080 and 8443 are used on the service endpoints backend, urbit planet pier itself, and the frontend gateway supports both - if people visit 8080 or 8443 it just works. So redirects just work, in spite of the mismatch. I think that was a bit harder than I thought due to NAT hairpinning issues which I worked around by moving to a different subnet that I had laying around, because I'm extra.

I unfortunately did not see the certificate renewed, because I gave up and reset `%acme` before realizing that NAT hairpinning was causing my problem. But I fortunately do have the network set up such that it should work next time around, and I can report back if it worked.

I'll remember to come back here and update in 60 days (in March or April) when the certificate is renewing!
